Firestore/core/src/firebase/firestore/api/firestore.mm

```diff
@@ -0,0 +1,183 @@
+#include "Firestore/core/src/firebase/firestore/util/executor_libdispatch.h"
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+#include "absl/memory/memory.h"
+
+namespace firebase {
+namespace firestore {
+namespace api {
+
+using firebase::firestore::api::Firestore;
+using firebase::firestore::auth::CredentialsProvider;
+using firebase::firestore::core::DatabaseInfo;
+using firebase::firestore::core::Transaction;
+using firebase::firestore::model::DocumentKey;
+using firebase::firestore::model::ResourcePath;
+using util::AsyncQueue;
+using util::Executor;
+using util::ExecutorLibdispatch;
+
+Firestore::Firestore(std::string project_id,
+                     std::string database,
+                     std::string persistence_key,
+                     std::unique_ptr<CredentialsProvider> credentials_provider,
+                     std::unique_ptr<AsyncQueue> worker_queue)
+    : database_id_{std::move(project_id), std::move(database)},
+      credentials_provider_{std::move(credentials_provider)},
+      persistence_key_{std::move(persistence_key)},
+      worker_queue_{std::move(worker_queue)} {
+  settings_ = [[FIRFirestoreSettings alloc] init];
+}
+
+AsyncQueue* Firestore::worker_queue() {
+  return [client_ workerQueue];
+}
+
+FIRFirestoreSettings* Firestore::settings() const {
+  std::lock_guard<std::mutex> lock{mutex_};
+  // Disallow mutation of our internal settings
+  return [settings_ copy];
+}
+
+void Firestore::set_settings(FIRFirestoreSettings* settings) {
+  std::lock_guard<std::mutex> lock{mutex_};
+  // As a special exception, don't throw if the same settings are passed
+  // repeatedly. This should make it more friendly to create a Firestore
+  // instance.
+  if (client_ && ![settings_ isEqual:settings]) {
+    HARD_FAIL(
+        "Firestore instance has already been started and its settings can "
+        "no longer be changed. You can only set settings before calling any "
+        "other methods on a Firestore instance.");
+  }
+  settings_ = [settings copy];
+}
+
+FIRCollectionReference* Firestore::GetCollection(
+    absl::string_view collection_path, FIRFirestore* firestore) {
+  EnsureClientConfigured();
+  ResourcePath path = ResourcePath::FromString(collection_path);
+  return [FIRCollectionReference referenceWithPath:path firestore:firestore];
+}
+
+DocumentReference Firestore::GetDocument(absl::string_view document_path) {
+  EnsureClientConfigured();
+  return DocumentReference{this, ResourcePath::FromString(document_path)};
+}
+
+FIRWriteBatch* Firestore::GetBatch(FIRFirestore* firestore) {
+  EnsureClientConfigured();
+  return [FIRWriteBatch writeBatchWithFirestore:firestore];
+}
+
+void Firestore::RunTransaction(TransactionBlock update_block,
+                               dispatch_queue_t queue,
+                               ResultOrErrorCompletion completion,
+                               FIRFirestore* firestore) {
+  EnsureClientConfigured();
```

Comment:
This check was implicit in Objective-C, and I spent quite a bit of time figuring out what was wrong.